rskj-core/src/main/java/org/ethereum/datasource/CacheSnapshotHandler.java

```diff
@@ -0,0 +1,123 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.ethereum.datasource;
+
+import org.ethereum.db.ByteArrayWrapper;
+import org.ethereum.util.MapSnapshot;
+import org.ethereum.util.TempFileCreator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import java.io.*;
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+public class CacheSnapshotHandler {
+
+    private static final Logger logger = LoggerFactory.getLogger(CacheSnapshotHandler.class);
+
+    private final MapSnapshot.Factory mapSnapshotFactory;
+    private final TempFileCreator tempFileCreator;
+
+    private final Path cacheSnapshotPath;
+
+    public CacheSnapshotHandler(@Nonnull Path cacheSnapshotPath) {
+        this(new MapSnapshot.Factory() {}, File::createTempFile, cacheSnapshotPath);
+    }
+
+    public CacheSnapshotHandler(@Nonnull MapSnapshot.Factory mapSnapshotFactory,
+                                @Nonnull TempFileCreator tempFileCreator,
+                                @Nonnull Path cacheSnapshotPath) {
+        this.mapSnapshotFactory = Objects.requireNonNull(mapSnapshotFactory);
+        this.tempFileCreator = Objects.requireNonNull(tempFileCreator);
+        this.cacheSnapshotPath = Objects.requireNonNull(cacheSnapshotPath);
+    }
+
+    public void load(@Nonnull Map<ByteArrayWrapper, byte[]> cache) {
+        File cacheSnapshotFile = cacheSnapshotPath.toFile();
+        String relativePath = Optional.ofNullable(cacheSnapshotFile.getParentFile())
+                .map(File::getName)
+                .orElse("") + "/" + cacheSnapshotFile.getName();
+
+        if (!cacheSnapshotFile.exists()) {
+            logger.info("Cache snapshot file '{}' not found. Returning empty cache", relativePath);
+            return;
+        }
+
+        try (MapSnapshot.In inSnapshot = mapSnapshotFactory.makeInputSnapshot(new BufferedInputStream(new FileInputStream(cacheSnapshotFile)))) {
+            inSnapshot.read(cache);
+
+            logger.info("Loaded {} cache entries from '{}'", cache.size(), relativePath);
+        } catch (IOException e) {
+            cache.clear();
+            File errFile = cacheSnapshotPath.resolveSibling(cacheSnapshotPath.getFileName() + ".err").toFile();
+            if (!errFile.exists()) {
+                if (!cacheSnapshotFile.renameTo(errFile)) {
+                    logger.error("Cannot rename cache snapshot file", e);
+                }
+            }
+            logger.error("Cannot read from cache snapshot file", e);
+        }
+    }
+
+    public void save(@Nonnull Map<ByteArrayWrapper, byte[]> cache) {
```

Comment:
I think we can explain here in a javadoc what's happening with the file (why and when is it being renamed and deleted, only a brief explanation).

Reply:
AMAZING!